### ISSUE TITLE
CI: pathogen-repo-build (aws-batch) does not output full logs to build.log

### ISSUE THREAD

**Issue by joverlee521** [nextstrain/.github#77]
## Context
First noticed when verifying tests for https://github.com/nextstrain/.github/pull/76. 

The CI workflow test for the `pathogen-repo-build` workflow using the `aws-batch` runtime does not output full logs to `build.log`. Most importantly, the line containing `AWS Batch Job ID` is not included in the `build.log` file so the workflow is unable to use the `AWS_BATCH_JOB_ID` to create the AWS Batch job summary. 

You can see the [job id in the logs](https://github.com/nextstrain/.github/actions/runs/8210709255/job/22458631526#step:8:40) via the GitHub Action UI. However, the workflow sets an [empty `AWS_BATCH_JOB_ID` envvar](https://github.com/nextstrain/.github/actions/runs/8210709255/job/22458631526?pr=76#step:9:14) and the `build.log` file of the [artifact](https://github.com/nextstrain/.github/actions/runs/8210709255/artifacts/1311002405) is also truncated.

---
## Possible solution

1. I tried using `tee -a build.log` in https://github.com/nextstrain/.github/commit/a8d9f4eccbfde5874a41c5ddd8336062a7abe1f1, but the `build.log` file of the [artifact](https://github.com/nextstrain/.github/actions/runs/8209746229/artifacts/1310798412) was still truncated. 

2. I thought maybe setting `PYTHONUNBUFFERED` in https://github.com/nextstrain/.github/commit/6f69e717b75f6e4c44e494cfe37bc104c4315661 would help, but the workflow is now using the standalone CLI so no dice. 

There must be some sort of stdout buffering that is getting lost and not output to the `build.log` file, but I'm out of ideas for now.

**Comment by victorlin** [comment]
The build.log from the two artifacts you linked to are rather different, but it does look like some weird truncation.

[2024-03-08](https://github.com/nextstrain/.github/actions/runs/8209746229/artifacts/1310798412):

```
$ hexdump -vC 03-08-build.log | head
00000000  2b 2b 20 6e 65 78 74 73  74 72 61 69 6e 20 62 75  |++ nextstrain bu|
00000010  69 6c 64 20 2d 2d 65 6e  76 20 42 4f 47 55 53 5f  |ild --env BOGUS_|
00000020  45 4e 56 56 41 52 20 2e  0a 5b 62 61 74 63 68 5d  |ENVVAR ..[batch]|
00000030  20 5b 32 30 32 34 2d 30  33 2d 30 38 54 32 32 3a  | [2024-03-08T22:|
00000040  32 38 3a 34 36 2b 30 30  3a 30 30 5d 20 09 53 61  |28:46+00:00] .Sa|
00000050  67 75 6c 65 6e 6b 6f 20  65 74 20 61 6c 2e 20 54  |gulenko et al. T|
00000060  72 65 65 54 69 6d 65 3a  20 4d 61 78 69 6d 75 6d  |reeTime: Maximum|
00000070  2d 6c 69 6b 65 6c 69 68  6f 6f 64 20 70 68 79 6c  |-likelihood phyl|
00000080  6f 64 79 6e 61 6d 69 63  20 61 6e 61 6c 79 73 69  |odynamic analysi|
00000090  73 0a 5b 62 61 74 63 68  5d 20 5b 32 30 32 34 2d  |s.[batch] [2024-|
```

[2024-03-09](https://github.com/nextstrain/.github/actions/runs/8210709255/artifacts/1311002405):

```
$ hexdump -vC 03-09-build.log | head
00000000  2b 2b 20 6e 65 78 74 73  74 72 61 69 6e 20 62 75  |++ nextstrain bu|
00000010  69 6c 64 20 2d 2d 65 6e  76 20 42 4f 47 55 53 5f  |ild --env BOGUS_|
00000020  45 4e 56 56 41 52 20 2e  0a 00 00 00 00 00 00 00  |ENVVAR .........|
00000030  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000040  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000050  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000060  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000070  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000080  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000090  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
```

A brief search pointed to `stdbuf` and `unbuffer` as potential solutions.

- [StackOverflow: Modifying "... | tee -a out.txt" to stream output live, rather than on completion?](https://stackoverflow.com/q/42938106)
- [StackOverflow: Force line-buffering of stdout in a pipeline](https://stackoverflow.com/q/11337041)

**Comment by joverlee521** [comment]
Thanks for the pointers @victorlin! 

Looks like [stdbuf has no effect on Python](https://stackoverflow.com/questions/55654364/why-stdbuf-has-no-effect-on-python), but I did try `unbuffer` on the [fix-ci-aws-batch branch](https://github.com/nextstrain/.github/compare/master...fix-ci-aws-batch). I manually triggered the [CI workflow](https://github.com/nextstrain/.github/actions/runs/8254001062) and unfortunately the build.log from the [artifact](https://github.com/nextstrain/.github/actions/runs/8254001062/artifacts/1319747001) is still truncated (although now truncated in a different way). 

This prompted me to dig into the Nextstrain CLI a little bit. I might be grasping at straws here, but could it have something to do with the [LogWatcher](https://github.com/nextstrain/cli/blob/4fa53169059047454fb10b55a584395239b61a1a/nextstrain/cli/runner/aws_batch/logs.py#L40) being a thread?

**Comment by victorlin** [comment]
At first I thought this was a problem with `tee` since the gitHub Actions logs show the full output. I tested that theory by redirecting the output to `build.log` directly ([CI run](https://github.com/nextstrain/.github/actions/runs/8255415288)). That still resulted in truncation, indicating that this is not a problem with `tee` but the act of saving the output to a file.

I tried an [alternative method](https://stackoverflow.com/a/3819814) to write output to a file while streaming it to console ([CI run](https://github.com/nextstrain/.github/actions/runs/8255330655)). While that did not work, the output in the console is a useful hint:

```
++ nextstrain build --env BOGUS_ENVVAR .
Nextstrain Run ID: 76b0b2cd-3a77-4641-aa7c-f81a650e14b2
Uploading /home/runner/work/.github/.github to S3
…
[batch] [2024-03-12T20:31:53+00:00] Inferred ancestral sequence states using TreeTime:
tail: build.log: file truncated
++ nextstrain build --env BOGUS_ENVVAR .
[batch] [2024-03-12T20:31:53+00:00] 	Sagulenko et al. TreeTime: Maximum-likelihood phylodynamic analysis
[batch] [2024-03-12T20:31:53+00:00] 	Virus Evolution, vol 4, https://academic.oup.com/ve/article/4/1/vex042/4794731
…
```

The line with `file truncated` is right where the truncation is observed in the downloaded `build.log`. I can think of these possibilities to explain such behavior:

1. Two processes are writing to `build.log` at the same time, without appending.
2. The process that writes to `build.log` is not handling output buffering properly.

I checked the CLI code and don't see any indication of (1). This leaves (2) unless there's other scenarios I've missed. Maybe a solution is to use `flush=True` in [`print_job_log()`](https://github.com/nextstrain/cli/blob/4fa53169059047454fb10b55a584395239b61a1a/nextstrain/cli/runner/aws_batch/__init__.py#L525)?

**Comment by joverlee521** [comment]
I added `flush=True` to `print_job_lob()` in https://github.com/nextstrain/cli/commit/a1a4946406ad47dd0a6a3b2ef0724bc1668c6d84 and manually ran the [Nextstrain CLI CI workflow](https://github.com/nextstrain/cli/actions/runs/8270546095) to create a test standalone. Then, I updated the version for the Nextstrain CLI in the `pathogen-repo-build` workflow to use the test standalone in https://github.com/nextstrain/.github/commit/082ca5378a11450a66241ae81c884f8460f7c8ea and manually ran the [CI workflow](https://github.com/nextstrain/.github/actions/runs/8270838105). 

The build.log is even more truncated in the [artifact](https://github.com/nextstrain/.github/actions/runs/8270838105/artifacts/1323758791) and only includes the post-completion logs.

**Comment by joverlee521** [comment]
The workflow is overwriting the `build.log` when we download and unzip the build results from AWS Batch 🤦‍♀️ 

So I took a closer look at yesterday's truncated `build.log` with fresh eyes and noticed the line 
```
unzipping: /home/runner/work/.github/.github/build.log
```
This made me realize the `build.log` is getting overwritten!

When we launch the build via aws-batch, the `build.log` file already exists and gets uploaded with the pathogen repo contents. I verified the `build.log` that is getting uploaded to AWS Batch locally with 
```
$ nextstrain build --aws-batch --attach 80dbb0d6-1348-4661-bbe1-01ed47290e31 .
$ cat build.log
++ nextstrain build --env BOGUS_ENVVAR .
```

This single-line file replaces the local `build.log` on the GH Action machine, then the post-job logs get appended and that is the `build.log` that we see in the [artifact](https://github.com/nextstrain/.github/actions/runs/8270838105/artifacts/1323758791).
